### ISSUE TITLE
Optimise `validation.py`

### ISSUE THREAD

**Issue by JoeZiminski** [neuroinformatics-unit/datashuttle#355]
Currently the validation is not well optimised and loops over the same list multiple times in various places. In general this may not be a problem as the lists arn't going to be huge (e.g. not more than 100 subjects I'd imagine). Nonetheless things might get slow for large projects, also it in some cases makes the logic hard to follow.

See `validate_names_against_project()` docstring, [this comment](https://github.com/neuroinformatics-unit/datashuttle/pull/349#issuecomment-2043408313) and docstring of `validate_list_of_names()` for discussions.

**Comment by aakash-test7** [comment]
@JoeZiminski is the optimisation issue still open . If yes should I look into it ?

**Comment by JoeZiminski** [comment]
Thanks @aakash-test7, this is completely in #464 which is just going through review, cheers though!
